impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java

```diff
@@ -163,65 +172,112 @@ public static boolean supportsConstraint(Class<? extends Annotation> annotation)
+        var useSasl = false;
+        var saslUsers = Optional.<Map<String, String>> empty();
+        var deprecatedSaslUsers = Optional.<Map<String, String>> empty();
+
         for (Annotation annotation : annotations) {
             if (annotation instanceof BrokerCluster brokerCluster) {
                 builder.brokersNum(brokerCluster.numBrokers());
             }
-            if (annotation instanceof KRaftCluster kRaftCluster) {
+            else if (annotation instanceof KRaftCluster kRaftCluster) {
                 builder.kraftMode(true);
                 builder.kraftControllers(kRaftCluster.numControllers());
             }
-            if (annotation instanceof ZooKeeperCluster) {
+            else if (annotation instanceof ZooKeeperCluster) {
                 builder.kraftMode(false);
             }
-            if (annotation instanceof Tls) {
+            else if (annotation instanceof Tls) {
                 tls = true;
-                try {
-                    builder.brokerKeytoolCertificateGenerator(new KeytoolCertificateGenerator());
-                }
-                catch (IOException e) {
-                    throw new RuntimeException(e);
-                }
-            }
-            if (annotation instanceof SaslPlainAuth.List saslPlainAuthList) {
-                builder.saslMechanism(PLAIN_SASL_MECHANISM_NAME);
-                sasl = true;
-                Map<String, String> users = new HashMap<>();
-                for (var user : saslPlainAuthList.value()) {
-                    users.put(user.user(), user.password());
-                }
-                builder.users(users);
+                processTls(builder::brokerKeytoolCertificateGenerator);
             }
-            else if (annotation instanceof SaslPlainAuth saslPlainAuth) {
-                builder.saslMechanism(PLAIN_SASL_MECHANISM_NAME);
-                sasl = true;
-                builder.users(Map.of(saslPlainAuth.user(), saslPlainAuth.password()));
-            }
-            if (annotation instanceof ClusterId clusterId) {
+            else if (annotation instanceof ClusterId clusterId) {
                 builder.kafkaKraftClusterId(clusterId.value());
             }
-            if (annotation instanceof Version version) {
+            else if (annotation instanceof Version version) {
                 builder.kafkaVersion(version.value());
             }
-            if (annotation instanceof BrokerConfig.List brokerConfigList) {
-                for (var config : brokerConfigList.value()) {
-                    builder.brokerConfig(config.name(), config.value());
-                }
+            else if (annotation instanceof SaslMechanism mechanism) {
+                useSasl = true;
+                builder.saslMechanism(Optional.ofNullable(mechanism.value()).orElse(PLAIN_SASL_MECHANISM_NAME));
+                var principals = Optional.ofNullable(mechanism.principals()).orElse(new SaslMechanism.Principal[]{});
```

Comment:
Should this be `orElseThrow` as there seems little point in setting up a cluster with no users? Actually answer my own question there will be a programmatic mechanism to add users to the cluster. 

Maybe we should log a warning if there are no configured principals so thats its clear to users that they need to do something else.

Reply:
Yes, I wondered about this.   You are correct, in the SCRAM case the user can can users later and that fine.  For the PLAIN, that's not the case.    The API could enforce the latter, or like you say, log an exception.  However this would be a job for the underlying KafkaCluster API rather than the annotation facade.  I'll do this separately.